Add Beginner Tutorials to the GETTSIM Documentation.

amageh opened the pull request:
### What problem do you want to solve?

We (David, Tim, & Annica) will use this PR to work on a set of four tutorials for the documentation that are targeted mainly towards beginner's first contact with GETTSIM. The tutorials build on and expand the existing interface tutorial. The following four tutorials are planned:

1. *Basic Interface*: Will be very similar to the existing interface tutorial but includes some adjustments to align it coherently with the following tutorials. Includes a minimal example of a possible input and use case.
2. *Advanced Interface*: Introduces larger simulated datasets to showcase the different combinations of input and target variables that GETTSIM supports.
3. *Parameters*: Showcases how parameters in `params_dict` can be changed  to make changes to the policy environment.
4. *Policy Functions*: Showcases how policy functions can be overridden to make changes to the policy environment.

Feedback is very welcome!

### Todo

- [x] Target the right branch and pick an appropriate title.
- [x] Tutorial Basic Interface
- [x] Tutorial Advanced Interface
- [x] Tutorial Parameters
- [x] Tutorial Policy Functions
- [ ] If it is a relevant contribution, put it in `CHANGES.rst`.
- [ ] Any steps that still need to be done.

amageh commented:
These are our drafts for the four tutorials listed in the PR description. The documentation build for this PR can be found [here](https://gettsim--232.org.readthedocs.build/en/232/tutorials/index.html). We have been working on the status of the `master` branch version of GETTSIM, i.e. the conda installation is behind.

Some questions that are still open:

- We generally had some troubles in coming up with suitable datasets and ended up using some from the lectures. We are happy for other suggestions. I think in our meetup a while ago, there was some dicussion about creating a simulated dataset (or function that simulates a dataset) for various purposes. Such a function would probably be quite useful.
- We are not very confident in the way inputs and outputs are explained. In the tutorials, we refer to the crosswalk page in the documentation but this seems to not be completely up to date? (i.e. there is a note in the basic tutorial on this list being somewhat wrong, that still needs to be removed) 
- In the `policy_functions` tutorial we 'overwrite' `arbeitsl_geld_2_m_hh` but this is actually not a key in the `policy_functions` dictionary before. The implementation seems to work anyway but we are not quite sure why (and think we should probably specify this more clearly).

We are happy to receive comments and feedback!

ChristianZimpelmann reviewed (changes requested):
Hi, 

looks altogether really good. Thanks for the great work! 

It was really helpful to get an overview. E.g. I really liked the references between the tutorials.

I made some in-line comments. And here some general remarks:

- Some tutorials could describe more what is done: Next we xxx... 
- The function tutorial explains much more basic Python stuff compared to the other tutorials (like what a #  does). This could be harmonized over the tutorials. I tend to like it if some stuff is explained.
- Just some idea: it would be nice if explanatory texts about GETTSIM or the German tax transfer system could be visually separated from the other text which explains the workflow of the tutorial. Not a problem if you don't find a solution that makes sense.

amageh commented on docs/tutorials/basic_interface.ipynb:
```diff
@@ -0,0 +1,837 @@
+    }
+   ],
+   "source": [
+    "policy_params, policy_functions = set_up_policy_environment(2020)"
```
Good point, I added this!

amageh commented on docs/tutorials/basic_interface.ipynb:
```diff
@@ -0,0 +1,837 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "These keys can be used to extract the exact parametrizations for a given policy group. The example below for instance shows the parameters that concern social security."
```
Unfortunately I am not aware of such a place, unless we link to the yaml files directly: https://github.com/iza-institute-of-labor-economics/gettsim/tree/master/gettsim/data ? Do you think this would make sense?

ChristianZimpelmann replied:
Ok, probably not. Then we can just leave it us is right now.

amageh commented on docs/tutorials/advanced_interface.ipynb:
```diff
@@ -0,0 +1,2544 @@
+       "(function(root) {\n",
+       "  function embed_document(root) {\n",
+       "    \n",
+       "  var docs_json = {\"3284b090-1e33-4f50-8abc-c77f35362ea5\":{\"roots\":{\"references\":[{\"attributes\":{\"center\":[{\"id\":\"1036\"},{\"id\":\"1039\"},{\"id\":\"1042\"},{\"id\":\"1045\"},{\"id\":\"1048\"},{\"id\":\"1051\"},{\"id\":\"1054\"},{\"id\":\"1057\"},{\"id\":\"1060\"},{\"id\":\"1063\"},{\"id\":\"1066\"},{\"id\":\"1069\"},{\"id\":\"1072\"},{\"id\":\"1075\"},{\"id\":\"1078\"},{\"i ...
```
Thanks, changed it!

amageh commented on docs/tutorials/advanced_interface.ipynb:
```diff
@@ -0,0 +1,2544 @@
+       "(function(root) {\n",
+       "  function embed_document(root) {\n",
+       "    \n",
+       "  var docs_json = {\"3284b090-1e33-4f50-8abc-c77f35362ea5\":{\"roots\":{\"references\":[{\"attributes\":{\"center\":[{\"id\":\"1036\"},{\"id\":\"1039\"},{\"id\":\"1042\"},{\"id\":\"1045\"},{\"id\":\"1048\"},{\"id\":\"1051\"},{\"id\":\"1054\"},{\"id\":\"1057\"},{\"id\":\"1060\"},{\"id\":\"1063\"},{\"id\":\"1066\"},{\"id\":\"1069\"},{\"id\":\"1072\"},{\"id\":\"1075\"},{\"id\":\"1078\"},{\"i ...
```
Corrected this one as well, thank you!

amageh commented on docs/tutorials/advanced_interface.ipynb:
```diff
@@ -0,0 +1,2544 @@
+{
```
I tried to revise it a bit, hopefully for the better. Please feel free to point out sections that still feel clunky/not well worded to you.

amageh commented:
Thank you very much for the review @ChristianZimpelmann! We tried to incorporate the feedback as much as possible. I commented on points that are still open and resolved those that are done. 

We generally tried to improve the wording a bit and added more comments to the code to make it easier to understand.

Regarding the visual separation, we dabbled with various options like indenting text or adding [alert-info-boxes](https://nbsphinx.readthedocs.io/en/0.2.7/markdown-cells.html). None of them seemed satisfactory in our opinion though, since the information in the text is pretty intertwined. Introducing such visual elements would probably require restructuring the current text quite a bit or make the notebooks rather messy.

amageh commented on docs/tutorials/policy_functions.ipynb:
```diff
@@ -0,0 +1,775 @@
+    "\n",
+    "**Context of the example:** *In the German system, there are some transfers for low-income families that cannot be received in combination. Per default, GETTSIM will always choose the most favorable transfers and set other transfers to zero. This assumption could model the behavior of households/families in a wrong way, if they do not always choose the optimal transfers (from a monetary pers ...
+    "\n",
+    "*To account for these frictions, we can turn off this aspect of GETTSIM so that we see all the transfers a family/household is entitled to, even if the transfers cannot be received in combination. This can be useful for further analysis. For example you could speculate which transfers Germans receive in reality and implement this in GETTSIM.*\n",
```
We discussed this point quite a bit to find a better example but couldn't settle on one that is as straightforward and easy to understand as this one (without changing the entire tutorial), as the current one just requires commenting out code of the existing policy function. There are probably better options that we might have missed though...

ChristianZimpelmann replied:
Then it's fine

ChristianZimpelmann commented:
> Thank you very much for the review @ChristianZimpelmann! We tried to incorporate the feedback as much as possible. I commented on points that are still open and resolved those that are done.
> 
> We generally tried to improve the wording a bit and added more comments to the code to make it easier to understand.
> 
> Regarding the visual separation, we dabbled with various options like indenting text or adding [alert-info-boxes](https://nbsphinx.readthedocs.io/en/0.2.7/markdown-cells.html). None of them seemed satisfactory in our opinion though, since the information in the text is pretty intertwined. Introducing such visual elements would probably require restructuring the current text quite a bit or make the notebooks rather messy.
> 
> One other remark: Currently, the basic tutorial and already [existing interface tutorial](https://gettsim.readthedocs.io/en/stable/tutorials/interface.html) are both still parts of the documentation. I am not sure whether it makes sense to keep both. We initially designed the basic and advanced interface tutorial to replace/expand the existing one but we didn't want to touch any existing materials in the documentation without consent, so I am mentioning it here.

Great! Thank you very much! I agree that the existing interface tutorial can be removed. What do you mean by basic tutorial?

amageh commented:
> Great! Thank you very much! I agree that the existing interface tutorial can be removed. What do you mean by basic tutorial?

I just meant the basic interface tutorial we added. It ended up being a very similar (just slightly more basic) version of the existing interface tutorial.

ChristianZimpelmann commented:
> > Great! Thank you very much! I agree that the existing interface tutorial can be removed. What do you mean by basic tutorial?
> 
> I just meant the basic interface tutorial we added. It ended up being a very similar (just slightly more basic) version of the existing interface tutorial.

Ok, I see! I understood that there are two duplicates.

From my side, it can be merged now.

hmgaudecker commented:
Thanks for all your work, looks great! 

Just a couple of very high-level comments:

- The installation is very short and it is not a tutorial. I think it should just go on the landing page, as a subsection? In the installation re conda: "please follow the installation instructions of its user guide." -> ""please follow conda's installation instructions."
- We have a section "Tutorials", then we do not need "Tutorial" in the subsection headers.
- I'd suggest avoiding the word "interface". Does not mean much to most people who will need the tutorials. Maybe 
  - Basic Usage
  - Advanced Usage
  - Changing Parameters of the Taxes and Transfers System
  - Changing Functions of the Taxes and Transfers System
  - Visualizing the taxes and transfers system **(plural!)**
  - Troubleshooting
  Please watch out for consistent Titlecase / lowercase everywhere, I do not have a strong opinion other than being consistent.
- Parameters: "reformed parameter" does not sound like correct English to me. And it is about the whole system including everything else. 
- Functions: does it help to add ";" to the end of the plot_dag() lines so the plot id's are not shown?
- Visualisation: Is this still a Tutorial or already a How-To Guide now that some of it is in the Advanced Usage? I guess one could make a case for both, given the imbalance I'd have a preference for How-To ATM.
- Debugging / Debug mode with an exception: Example needs a cross reference to a place where this construct of user functions is explained. Most users will not understand what is going on in the example. Also, the description of which columns are returned could be more verbose, e.g., why it is useful? (you will be able to see all inputs to the function(s) that fail.

amageh commented:
Thank you very much for the feedback @hmgaudecker ! I think we have addressed all of the comments (feel free to point out if we have missed anything). 

One last open point concerns the debgugging/troubleshooting tutorial which we hadn't previously worked on. I'm not sure why, but the debug mode, or rather the example we have in the tutorial, does not seem to be throwing the warning or returning only a subset of variables anymore. Maybe someone who has more knowledge of the internals and recent changes than me can weigh in  @MaxBlesch @ChristianZimpelmann ?

I have pushed [my run](https://gettsim--232.org.readthedocs.build/en/232/tutorials/debugging.html) of the tutorial (I updated some function arguments since they have changed since the tutorial was published) now, so you can compare to the old version [here](https://gettsim.readthedocs.io/en/stable/tutorials/debugging.html) in the main documentation.

amageh commented:
Nvm my last comment, it's working. Thanks @MaxBlesch !

hmgaudecker reviewed (commented):
Looks great, thank you! 

Just out of scope of this one, but I think we should keep it in there anyhow is what to do with the crosswalk. That one should not be placed as prominently anymore, if we want to keep it in there, we want to hide it in some section "compatibility" or so, but I do not have great ideas at present.

https://gettsim--232.org.readthedocs.build/en/232/functions.html# sort of replaces it for the output columns, that we should document a bit better. Maybe have a page with 

Data columns
- Basic user inputs
- Typical outputs
- All potential outputs

@MaxBlesch, could you make a suggestion on that one? Feel free to just open a new issue and/or PR if you think that will grow too large (having written this, it already feels like that), I do not have enough bandwidth these days to fully think through it.

MaxBlesch commented:
Yes. I already work on something like that. 
https://gettsim--230.org.readthedocs.build/en/230/variables.html has all basic inputs. I had the idea to extend that here to our default [targets](https://github.com/iza-institute-of-labor-economics/gettsim/blob/db3d0392d94bde7d62cb12ca9fd19c9f35cf5902/gettsim/config.py#L40).

But all potential outputs are given by the documentation of functions. Once there are complete docstrings. That's the documentation we want to have?!

In any case out of scope of this PR. Very nice work!

hmgaudecker commented:
> https://gettsim--230.org.readthedocs.build/en/230/variables.html has all basic inputs. I had the idea to extend that here to our default [targets](https://github.com/iza-institute-of-labor-economics/gettsim/blob/db3d0392d94bde7d62cb12ca9fd19c9f35cf5902/gettsim/config.py#L40).
> 
> But all potential outputs are given by the documentation of functions. Once there are complete docstrings. That's the documentation we want to have?!

Yes, excellent! In general, we will need think hard on how to package things, so that new users get the important bits at a glance without being overwhelmed by all the information there is in the background. 

> In any case out of scope of this PR. Very nice work!

Indeed, thanks @amageh, @davpahl, @Trichter33!
